[FLINK-32287][docs] Add doc for truncate table statement (#23046)
apache/flink: docs/content/docs/dev/table/sql/truncate.md

---
title: "TRUNCATE Statements"
weight: 8
type: docs
aliases:
- /dev/table/sql/truncate.html
---
<!--
Licensed to the Apache Software Foundation (ASF) under one
or more contributor license agreements.  See the NOTICE file
distributed with this work for additional information
regarding copyright ownership.  The ASF licenses this file
to you under the Apache License, Version 2.0 (the
"License"); you may not use this file except in compliance
with the License.  You may obtain a copy of the License at

  http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing,
software distributed under the License is distributed on an
"AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
KIND, either express or implied.  See the License for the
specific language governing permissions and limitations
under the License.
-->

# TRUNCATE Statements

> 

Reviewer: ```suggestion
{{< label Batch >}}
```
Don't forget to break a new line, you can follow other docs.

TRUNCATE statements are used to delete all rows from a table without dropping the table itself.

> <span class="label label-danger">Attention</span> Currently, `TRUNCATE` statement only supports in batch mode, and it requires the target table connector implements the  {{< gh_link file="flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/sink/abilities/SupportsTruncate.java" name="SupportsTruncate" >}}

Reviewer: nit: 
```suggestion
<span class="label label-danger">Attention</span> Currently, `TRUNCATE` statement only is supported in batch mode, and it requires the target table connector implements the  {{< gh_link file="flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/sink/abilities/SupportsTruncate.java" name="SupportsTruncate" >}}
```

> interface to support the row-level delete. An exception will be thrown if trying to `TRUNCATE` the table which has not implements the related interface.

Reviewer: ```suggestion
interface to support the row-level delete. An exception will be thrown if trying to `TRUNCATE` a table which has not implemented the related interface.
```

tEnv.executeSql("SELECT * FROM Orders").print();
// +--------------------------------+--------------------------------+-------------+
// |                           user |                        product |      amount |
// +--------------------------------+--------------------------------+-------------+
// |                           Lili |                          Apple |           1 |
// |                        Jessica |                         Banana |           2 |
// |                       Mr.White |                        Chicken |           3 |
// +--------------------------------+--------------------------------+-------------+
// 3 rows in set
        
// truncate the table "Orders"
tEnv.executeSql("TRUNCATE TABLE Orders").await();
tEnv.executeSql("SELECT * FROM Orders").print();
// Empty set
```
{{< /tab >}}
{{< tab "Scala" >}}
```scala
val env = StreamExecutionEnvironment.getExecutionEnvironment()
val settings = EnvironmentSettings.newInstance().inBatchMode().build()
val tEnv = StreamTableEnvironment.create(env, settings)

// register a table named "Orders"

> tEnv.executeSql("CREATE TABLE Orders (`user` STRING, product STRING, amount INT) WITH (...)");

Reviewer: ```suggestion
tEnv.executeSql("CREATE TABLE Orders (`user` STRING, product STRING, amount INT) WITH (...)")
```

// insert values

> tEnv.executeSql("INSERT INTO Orders VALUES ('Lili', 'Apple', 1), ('Jessica', 'Banana', 2), ('Mr.White', 'Chicken', 3)").await();

Reviewer: ```suggestion
tEnv.executeSql("INSERT INTO Orders VALUES ('Lili', 'Apple', 1), ('Jessica', 'Banana', 2), ('Mr.White', 'Chicken', 3)").await()
```

> tEnv.executeSql("SELECT * FROM Orders").print();

Reviewer: dito

// +--------------------------------+--------------------------------+-------------+
// |                           user |                        product |      amount |
// +--------------------------------+--------------------------------+-------------+
// |                           Lili |                          Apple |           1 |
// |                        Jessica |                         Banana |           2 |
// |                       Mr.White |                        Chicken |           3 |
// +--------------------------------+--------------------------------+-------------+
// 3 rows in set
// truncate the table "Orders"

> tEnv.executeSql("TRUNCATE TABLE Orders").await();

Reviewer: dito

> tEnv.executeSql("SELECT * FROM Orders").print();

Reviewer: dito

// Empty set
```
{{< /tab >}}
{{< tab "Python" >}}
```python
env_settings = EnvironmentSettings.in_batch_mode()
table_env = TableEnvironment.create(env_settings)

# register a table named "Orders"

> table_env.executeSql("CREATE TABLE Orders (`user` STRING, product STRING, amount INT) WITH (...)")

Reviewer: ```suggestion
table_env.execute_sql("CREATE TABLE Orders (`user` STRING, product STRING, amount INT) WITH (...)")
```

> table_env.executeSql("CREATE TABLE Orders (`user` STRING, product STRING, amount INT) WITH (...)");

Reviewer: ```suggestion
table_env.execute_sql("CREATE TABLE Orders (`user` STRING, product STRING, amount INT) WITH (...)")
```

# insert values

> table_env.executeSql("INSERT INTO Orders VALUES ('Lili', 'Apple', 1), ('Jessica', 'Banana', 2), ('Mr.White', 'Chicken', 3)").wait()

Reviewer: ```suggestion
table_env.execute_sql("INSERT INTO Orders VALUES ('Lili', 'Apple', 1), ('Jessica', 'Banana', 2), ('Mr.White', 'Chicken', 3)").wait()
```

> table_env.executeSql("INSERT INTO Orders VALUES ('Lili', 'Apple', 1), ('Jessica', 'Banana', 2), ('Mr.White', 'Chicken', 3)").wait();

Reviewer: dito

> table_env.executeSql("SELECT * FROM Orders").print();

Reviewer: dito

# +--------------------------------+--------------------------------+-------------+
# |                           user |                        product |      amount |
# +--------------------------------+--------------------------------+-------------+
# |                           Lili |                          Apple |           1 |
# |                        Jessica |                         Banana |           2 |
# |                       Mr.White |                        Chicken |           3 |
# +--------------------------------+--------------------------------+-------------+
# 3 rows in set
# truncate the table "Orders"

> table_env.executeSql("TRUNCATE TABLE Orders").wait()

Reviewer: ```suggestion
table_env.execute_sql("TRUNCATE TABLE Orders").wait()
```

> table_env.executeSql("TRUNCATE TABLE Orders").wait();

Reviewer: dito

> table_env.executeSql("SELECT * FROM Orders").print()

Reviewer: ```suggestion
table_env.execute_sql("SELECT * FROM Orders").print()
```

> table_env.executeSql("SELECT * FROM Orders").print();

Reviewer: dito

---
title: "TRUNCATE TABLE Statement"
weight: 8
type: docs
aliases:
- /dev/table/sql/truncate.html
---
<!--
Licensed to the Apache Software Foundation (ASF) under one
or more contributor license agreements.  See the NOTICE file
distributed with this work for additional information
regarding copyright ownership.  The ASF licenses this file
to you under the Apache License, Version 2.0 (the
"License"); you may not use this file except in compliance
with the License.  You may obtain a copy of the License at

  http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing,
software distributed under the License is distributed on an
"AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
KIND, either express or implied.  See the License for the
specific language governing permissions and limitations
under the License.
-->

> # TRUNCATE TABLE Statement

Reviewer: Change to `TRUNCATE Statements` just like [`ANALYZE Statements`](http://localhost:1313/docs/dev/table/sql/analyze/)?

TRUNCATE TABLE statement are used to delete all rows from a table without dropping the table itself.

> TRUNCATE TABLE statement is only supported in batch execution mode now.

Reviewer: dito

## Run a TRUNCATE TABLE statement

{{< tabs "truncate" >}}

> {{< tab "Java" >}}

Reviewer: Why miss `Scala`/`Python` example?

> TRUNCATE TABLE statement can be executed with the `executeSql()` method of the `TableEnvironment`. The `executeSql()` method returns nothing for a successful TRUNCATE TABLE operation, otherwise will throw an exception.

Reviewer: Actually, it doesn't return nothing. It returns `TABLE_RESULT_OK`. If we can't describe what is returned precisely , I'd like to not to mention it.

The following examples show how to run a TRUNCATE TABLE statement in `TableEnvironment`.
{{< /tab >}}
{{< tab "SQL CLI" >}}

TRUNCATE TABLE statement can be executed in [SQL CLI]({{< ref "docs/dev/table/sqlClient" >}}).

The following examples show how to run a TRUNCATE TABLE statement in SQL CLI.

{{< /tab >}}
{{< /tabs >}}

{{< tabs "a5de1760-e363-4b8d-9d6f-0bacb35b9dcf" >}}
{{< tab "Java" >}}
```java
TableEnvironment tableEnv = TableEnvironment.create(...);

// register a table named "Orders"
tableEnv.executeSql(
        "CREATE TABLE Orders (" +
        " `user` BIGINT NOT NULl comment 'this is primary key'," +
        " product VARCHAR(32)," +
        " amount INT," +
        " ts TIMESTAMP(3) comment 'notice: watermark'," +
        " ptime AS PROCTIME() comment 'this is a computed column'," +
        " PRIMARY KEY(`user`) NOT ENFORCED," +
        " WATERMARK FOR ts AS ts - INTERVAL '1' SECONDS" +
        ") with (...)");

> // print the schema

Reviewer: why `// print the schema`?

Reviewer: Also, please follow the examples in `delete statement` to make sure we can really delete all rows using truncate table statement.
